### PR TITLE
Support more formats for replay

### PR DESCRIPTION
**Pull request by gnegri**

### CODE REVIEW

File: common/reportMetrics.py [hunk 1]
```diff
@@ -369,108 +373,343 @@ def process_replay(file_path):
...
+                    # send final chunk
+                    if len(to_send_metric_data) != 0:
+                        send_data([to_send_metric_data, min_timestamp_epoch, max_timestamp_epoch], file_path, chunk_count + 1)
+                        chunk_count += 1
+                    logger.debug("Total chunks created: " + str(chunk_count))
                 save_grouping(grouping_map)
 
 
+def replay_sar(metric_file_path, grouping_map):
+    print('Replaying sar file')
```
Comment (moharnab123saikia) [inline common/reportMetrics.py]: use logger

File: common/reportMetrics.py [hunk 2]
```diff
@@ -369,108 +373,343 @@ def process_replay(file_path):
...
+        if len(to_send_metric_data) != 0:
+            send_data([to_send_metric_data, min_timestamp_epoch, max_timestamp_epoch], metric_file_path, chunk_count + 1)
+            chunk_count += 1
+        logger.debug("Total chunks created: " + str(chunk_count))
+        save_grouping(grouping_map)
+    return
+
+
+def replay_db2(log_file_path):
+    print('Replaying db2 file')
```
Comment (moharnab123saikia) [inline common/reportMetrics.py]: use logger

File: common/reportMetrics.py [hunk 3]
```diff
@@ -369,108 +373,343 @@ def process_replay(file_path):
...
+        last_key = line_arr[last_key_pos]
+        obj[last_key] = last_val
+        if last_key_pos == 0:
+            break
+        line_arr = line_arr[:last_key_pos]
+    return
+
+
+def replay_gpfs(log_file_path):
+    print('Replaying gpfs file')
```
Comment (moharnab123saikia) [inline common/reportMetrics.py]: logger

File: common/reportMetrics.py [hunk 4]
```diff
@@ -369,108 +373,343 @@ def process_replay(file_path):
...
+            if line_count == parameters['chunkLines']:
+                logger.debug("--- Chunk creation time: %s seconds ---" % (time.time() - start_time))
+                send_data(current_row, log_file_path, chunk_count)
+                current_row = []
+                chunk_count += 1
+                line_count = 0
+                start_time = time.time()
+            # build json entry
+            entry = dict()
+            entry['tag'] = 'gpfs-replay'
```
Comment (moharnab123saikia) [inline common/reportMetrics.py]: isn't there any host to get from the file? or can we just use localhost

File: common/reportMetrics.py [hunk 5]
```diff
@@ -369,108 +373,343 @@ def process_replay(file_path):
...
+        current_row = []
+        current_obj = dict()
+        start_time = time.time()
+        line = log_file.readline()
+        while line:
+            # empty line denotes new log entry. aggregate current string
+            if not line.strip():
+                # build json entry
+                entry = dict()
+                entry['tag'] = 'db2-replay'
```
Comment (moharnab123saikia) [inline common/reportMetrics.py]: is a hostname or localhost can be used

File: common/reportMetrics.py [hunk 6]
```diff
@@ -369,108 +373,343 @@ def process_replay(file_path):
...
 
 
+def replay_sar(metric_file_path, grouping_map):
+    print('Replaying sar file')
+    with open(metric_file_path) as metric_file:
+        line = metric_file.readline()
+        header = line.split()
+        instance = header[2][1:-1]
+        date = header[3]
+        _ = metric_file.readline()
```
Comment (moharnab123saikia) [inline common/reportMetrics.py]: the naming seems strange

File: common/reportMetrics.py [hunk 7]
```diff
@@ -369,108 +373,343 @@ def process_replay(file_path):
...
+                    if len(to_send_metric_data) != 0:
+                        send_data([to_send_metric_data, min_timestamp_epoch, max_timestamp_epoch], file_path, chunk_count + 1)
+                        chunk_count += 1
+                    logger.debug("Total chunks created: " + str(chunk_count))
                 save_grouping(grouping_map)
 
 
+def replay_sar(metric_file_path, grouping_map):
+    print('Replaying sar file')
+    with open(metric_file_path) as metric_file:
```
Comment (moharnab123saikia) [inline common/reportMetrics.py]: check if file exists

File: common/reportMetrics.py [hunk 8]
```diff
@@ -369,108 +373,343 @@ def process_replay(file_path):
...
+        obj[last_key] = last_val
+        if last_key_pos == 0:
+            break
+        line_arr = line_arr[:last_key_pos]
+    return
+
+
+def replay_gpfs(log_file_path):
+    print('Replaying gpfs file')
+    with open(log_file_path) as log_file:
```
Comment (moharnab123saikia) [inline common/reportMetrics.py]: check if file exists before opening

File: common/reportMetrics.py (hunk 7 above)
Comment (gnegri) [reply common/reportMetrics.py]: added check when input as parameter, already checked at top of tag that calls this

File: common/reportMetrics.py (hunk 8 above)
Comment (gnegri) [reply common/reportMetrics.py]: added check when input as parameter, already checked at top of tag that calls this

File: common/reportMetrics.py (hunk 4 above)
Comment (gnegri) [reply common/reportMetrics.py]: see updated tag. 

https://www.ibm.com/support/knowledgecenter/en/STXKQY_4.2.3/com.ibm.spectrum.scale.v4r23.doc/bl1pdg_errlogs.htm

> The GPFS log file is named mmfs.log.date.nodeName

File: common/reportMetrics.py (hunk 5 above)
Comment (gnegri) [reply common/reportMetrics.py]: Will now use hostname in log entry or localhost if unavailable

File: common/reportMetrics.py [hunk 9]
```diff
@@ -369,108 +373,354 @@ def process_replay(file_path):
...
+                line_count += 1
+                current_obj = dict()
+            else:
+                # first line starts with timestamp
+                if len(current_obj) == 0:
+                    timestamp_str = line.split()[0]
+                    current_obj['timestamp'] = _get_timestamp_db2(timestamp_str, parameters['timeZone'])
+                    current_obj['id'] = line.split()[1]
+                    current_obj['LEVEL'] = line[line.rfind(' '):]
+                elif line[:3] == 'ARG' or line[:4] == 'DATA' or line[:7] == 'MESSAGE':
```
[the remark this answers is not in the document]
Comment (gnegri) [reply common/reportMetrics.py]: won't cause errors with line lengths < 3

File: common/reportMetrics.py [hunk 10]
```diff
@@ -369,108 +373,354 @@ def process_replay(file_path):
...
+            timestamp = first_timestamp_epoch + int(agent_config_vars['samplingInterval']) * 60 * 1000 * row_count
+            current_row['timestamp'] = str(timestamp)
+            if min_timestamp_epoch == 0 or min_timestamp_epoch > long(timestamp):
+                min_timestamp_epoch = long(timestamp)
+            if max_timestamp_epoch == 0 or max_timestamp_epoch < long(timestamp):
+                max_timestamp_epoch = long(timestamp)
+
+            # metric values
+            for i in range(2, len(row)):
+                column_name = field_names[i].strip()
```
[the remark this answers is not in the document]
Comment (gnegri) [reply common/reportMetrics.py]: no errors if len(row) <=2, not a concern

File: common/reportMetrics.py [hunk 11]
```diff
@@ -369,108 +373,354 @@ def process_replay(file_path):
...
+            # Read each line from csv and generate a json
+            current_row = dict()
+            if current_line_count == parameters['chunkLines']:
+                send_data([to_send_metric_data, min_timestamp_epoch, max_timestamp_epoch], metric_file_path, chunk_count + 1)
+                to_send_metric_data = []
+                current_line_count = 0
+                chunk_count += 1
+
+            row = line.split()
+            if row[2:] == field_names[2:] or 'RESTART' in row or not row:
```
[the remark this answers is not in the document]
Comment (gnegri) [reply common/reportMetrics.py]: won't throw any sort of index out of bounds error. skipping an empty line is part of this filter, so this is just an earlier condition for that, essentially.
`
>>> row = [1]
>>> line = [3]
>>> print(row[2:] == line[2:])
True
`

File: common/reportMetrics.py [hunk 12]
```diff
@@ -369,108 +373,354 @@ def process_replay(file_path):
...
 
 
+def replay_sar(metric_file_path, grouping_map):
+    logger.info('Replaying sar file')
+    with open(metric_file_path) as metric_file:
+        line = metric_file.readline()
+        header = line.split()
+        instance = header[2][1:-1]
+        # assume DD/MM/YYYY
+        date = header[3]
```
[the remark this answers is not in the document]
Comment (gnegri) [reply common/reportMetrics.py]: an issue here would be indicative of an issue in file format, ie failure is the correct result

File: common/reportMetrics.py [hunk 13]
```diff
@@ -369,108 +373,354 @@ def process_replay(file_path):
...
+        format = '%m/%d/%Y %I:%M:%S %p'
+        _ = metric_file.readline()
+        metrics = metric_file.readline()
+        field_names = metrics.split()
```
[the remark this answers is not in the document]
Comment (gnegri) [reply common/reportMetrics.py]: an issue here would be indicative of an issue in file format, ie failure is the correct result

File: common/reportMetrics.py [hunk 14]
```diff
@@ -369,108 +373,354 @@ def process_replay(file_path):
...
+        current_row = []
+        current_obj = dict()
+        start_time = time.time()
+        line = log_file.readline()
+        while line:
+            # empty line denotes new log entry. aggregate current string
+            if not line.strip():
+                # build json entry
+                entry = dict()
+                entry['tag'] = current_obj['HOSTNAME'] or 'localhost'
```
[the remark this answers is not in the document]
Comment (gnegri) [reply common/reportMetrics.py]: update to check for key in name and non null key-value

**Pull request merged**